internal/client/client.go

```diff
@@ -478,41 +564,725 @@ func (c *Client[H, Hasher, N, E, Header]) Block(hash H) (*generic.SignedBlock[N,
+		return prepareStorageChangesResultDiscard{common.ImportResultUnknownParent{}}, nil
+	} else if _, ok := stateAction.(common.StateActionSkip); ok {
+		enactState = false
+		storageChanges = nil
+	} else if _, ok := stateAction.(common.StateActionExecute); ok {
+		enactState = true
+		storageChanges = nil
+	} else if _, ok := stateAction.(common.StateActionExecuteIfPossible); ok {
+		enactState = true
+		storageChanges = nil
+	}
+
+	var storageChangesToApply common.StorageChanges
+
+	if enactState && storageChanges != nil {
+		// we have storage changes and should enact the state, so we don't need to do anything here
+		storageChangesToApply = storageChanges
+	} else if enactState && storageChanges == nil && importBlock.Body != nil {
+		// We should enact state, but don't have any storage changes, so we need to execute the block
+		runtimeApi := c.RuntimeApi()
+
+		runtimeApi.SetCallContext(core.CallContextOnchain)
+		if c.config.EnableImportProofRecording {
+			runtimeApi.RecordProof()
+			recorder := runtimeApi.ProofRecorder()
+			if recorder == nil {
+				panic("Proof recording is enabled in the line above; qed.")
+			}
+
+			runtimeApi.RegisterExtension(recorder)
+		}
+
+		err := runtimeApi.ExecuteBlock(parentHash, generic.NewBlock[Hasher](importBlock.Header, importBlock.Body))
+		if err != nil {
+			return nil, err
+		}
+
+		state, err := c.backend.StateAt(parentHash)
+		if err != nil {
+			return nil, err
+		}
+
+		genStorageChanges, err := runtimeApi.IntoStorageChanges(state, parentHash)
+		if err != nil {
+			return nil, err
+		}
+
+		if importBlock.Header.StateRoot() != genStorageChanges.TransactionStorageRoot {
+			return nil, blockchain.ErrInvalidStateRoot
+		}
+
+		storageChangesToApply = common.Changes[H, Hasher](genStorageChanges)
+	} else if enactState && storageChanges == nil && importBlock.Body == nil {
+		// No block body, no storage changes
+		storageChangesToApply = nil
+	} else if !enactState {
+		// We should not enact the state, so we set the storage changes to None
+		storageChangesToApply = nil
+	}
+
+	return prepareStorageChangesResultImport{storageChangesToApply}, nil
+}
+
+func (c *Client[H, Hasher, N, E, Executor, Header, RA]) applyBlock(
+	operation *api.ClientImportOperation[H, Hasher, N, Header, E],
+	importBlock common.BlockImportParams[H, N, E, Header],
+	storageChanges common.StorageChanges,
+) (common.ImportResult, error) {
+	if len(importBlock.Intermediates) > 0 {
+		return nil, blockchain.ErrIncompletePipeline
+	}
+
+	if importBlock.ForkChoice == nil {
+		return nil, blockchain.ErrIncompletePipeline
+	}
+
+	var importHeaders PrePostHeaders[N, H, Header]
+
+	if len(importBlock.PostDigests) == 0 {
+		importHeaders = PrePostHeadersSame[N, H, Header]{importBlock.Header}
+	} else {
+		postHeader := importBlock.Header.Clone()
+		for _, item := range importBlock.PostDigests {
+			postHeader.DigestMut().Push(item)
+		}
+		importHeaders = PrePostHeadersDifferent[N, H, Header]{importBlock.Header, importBlock.Header}
```

Comment:
Shouldn't this be
```suggestion
		importHeaders = PrePostHeadersDifferent[N, H, Header]{importBlock.Header, postHeader}
```
?

Reply:
very good catch! thks!